### PR TITLE
feat!: Implement FEATURE_BUNDLE_1 RFC 0004

### PR DESCRIPTION
**Pull request by mwiebe**
### What was the problem/requirement? (What/Why)

[RFC 0004](https://github.com/OpenJobDescription/openjd-specifications/blob/mainline/rfcs/0004-enhanced-limits-and-capabilities.md) was accepted 1/9, and needs implementation in openjd-model, openjd-sessions, etc.

### What was the solution? (How)

Implement the extension in the model parsing. For the syntax sugar, this also includes a new method on the StepTemplate that de-sugars it into the syntax without the sugar.

### What is the impact of this change?

Progress towards this extension being available for use.

### How was this change tested?

This change introduces unit tests for the extension.

I used the following prompt for LLM evaluation of this code change against RFC 0004 in openjd-specifications:

> You are a helpful RFC conformance reviewer. Another agent has
> implemented RFC 0004, along with tests, in the most recent commit. Your
> job is to carefully inspect the RFC, the RFC-specific details in the
> OpenJD spec, and look at all the code changes to evaluate compliance.
> Produce a report on the quality of the implementation. How well does it
> conform? Give each relevant section/subsection of the spec a score. Also
> look at the unit tests and evaluate that it is covering all edge cases
> of the spec. Write your report to the file CONFORMANCE_REPORT.md.

All code also reviewed by hand.

### Was this change documented?

Extension documented in openjd-specifications, this change adds comments/docstrings as appropriate.

### Is this a breaking change?

Yes, in a subtle way with some fields that can now be FormatString instead of just int.

BREAKING CHANGE: The types of some model class fields are extended from
    Optional[int] to Optional[int | FormatString], so downstream code
    that relies on those types can fail type checking.

### Does this change impact security?

No.

----

*By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.*

### CODE REVIEW

File: src/openjd/model/v2023_09/_model.py [hunk 1]
```diff
@@ -285,7 +292,31 @@ class CancelationMethodNotifyThenTerminate(OpenJDModel_v2023_09):
...
+                # Try to parse as int, fail if not
+                try:
+                    return int(v)
+                except ValueError:
+                    raise ValueError(
+                        "notifyPeriodInSeconds as a format string requires the FEATURE_BUNDLE_1 extension."
+                    )
+            return validate_int_fmtstring_field(v, ge=1, context=context)
+        if isinstance(v, int):
+            if v < 1 or v > 600:
```
Comment (jericht) [inline src/openjd/model/v2023_09/_model.py:327]: Is this 600 max value new?

**Review by jericht — DISMISSED** [review]

File: test/openjd/model/test_parse.py [hunk 2]
```diff
@@ -194,13 +194,15 @@ class MockExtensionName(str, Enum):
...
     SUPPORTED_NAME = "SUPPORTED_NAME"
+    FEATURE_BUNDLE_1 = "FEATURE_BUNDLE_1"
 
 
 class MockExtensionNameWithTwoNames(str, Enum):
     """A mock enum with only SUPPORTED_NAME for testing."""
 
     SUPPORTED_NAME = "SUPPORTED_NAME"
     ANOTHER_SUPPORTED_NAME = "ANOTHER_SUPPORTED_NAME"
+    FEATURE_BUNDLE_1 = "FEATURE_BUNDLE_1"
```
Comment (epmog) [inline test/openjd/model/test_parse.py:205]: should these two enums be changed? Docstring indicates otherwise

File: src/openjd/model/v2023_09/_model.py [hunk 3]
```diff
@@ -187,17 +201,19 @@ class ValueReferenceConstants(Enum):
...
+    # because the template name can contain format strings
 
     def __new__(cls, value: str, *, context: ModelParsingContextInterface = ModelParsingContext()):
         return super().__new__(cls, value, context=context)
 
 
 JobName = Annotated[
     str,
-    StringConstraints(min_length=1, max_length=128, strict=True, pattern=_standard_string_regex),
+    StringConstraints(min_length=1, max_length=512, strict=True, pattern=_standard_string_regex),
```
Comment (sakshie95) [inline src/openjd/model/v2023_09/_model.py:213]: Have you kept this as a buffer cz of comment in the code above?

```
# Max length is validated after resolution in Job model, not here
# because the template name can contain format strings
```

File: src/openjd/model/v2023_09/_model.py [hunk 4]
```diff
@@ -415,12 +493,91 @@ class EmbeddedFileText(OpenJDModel_v2023_09):
...
+            raise ValueError(f"name must be at most {max_len} characters long")
+        return v
+
+    @field_validator("filename")
+    @classmethod
+    def _validate_filename(cls, v: Optional[Filename], info: ValidationInfo) -> Optional[Filename]:
+        if v is None:
+            return v
+        context = cast(Optional[ModelParsingContext], info.context)
+        max_len = 256 if context and "FEATURE_BUNDLE_1" in context.extensions else 64
```
Comment (sakshie95) [inline src/openjd/model/v2023_09/_model.py:511]: nit: maybe using constants for the constraints would be better here so its easy to update

File: src/openjd/model/v2023_09/_model.py [hunk 5]
```diff
@@ -1052,6 +1209,15 @@ class Environment(OpenJDModel_v2023_09):
 
     _template_variable_scope = ResolutionScope.SESSION
 
+    @field_validator("name")
+    @classmethod
+    def _validate_environment_name(cls, v: str, info: ValidationInfo) -> str:
+        context = cast(Optional[ModelParsingContext], info.context)
+        max_len = 512 if context and "FEATURE_BUNDLE_1" in context.extensions else 64
```
Comment (sakshie95) [inline src/openjd/model/v2023_09/_model.py:1216]: nit: Same as above, maybe using constants would be better here

File: src/openjd/model/v2023_09/_model.py (hunk 1 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:327]: It's not, you can see it in specified and then applied here:

* https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#53-cancelationmethod
* https://github.com/OpenJobDescription/openjd-model-for-python/blob/mainline/src/openjd/model/v2023_09/_model.py#L254

File: src/openjd/model/v2023_09/_model.py (hunk 3 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:213]: The bounds that we put in the type annotations always apply when validating/parsing an input with pydantic, so the upper limit has to be the maximum of upper limits supported. Later, it is dynamically checked against the extension-specific limit.

File: src/openjd/model/v2023_09/_model.py (hunk 4 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:511]: I think when you apply LLM code prompting, this convention that was true before may be reversed now - with the values directly here there are fewer indirections that are necessary to cross-validate against the written spec, for example.

File: src/openjd/model/v2023_09/_model.py [hunk 6]
```diff
@@ -398,13 +473,16 @@ class EmbeddedFileText(OpenJDModel_v2023_09):
             will have its execute-permissions set.
             Default: False
         data (FormatString): The text data to write to the file.
+        endOfLine (Optional[EndOfLine]): Line ending style. Requires FEATURE_BUNDLE_1 extension.
+            Default: AUTO
```
Comment (epmog) [inline src/openjd/model/v2023_09/_model.py:477]: do we need to describe AUTO? Is it based on operating system Windows vs. Posix? Or extension type? I don't see it listed in the RFC

File: src/openjd/model/v2023_09/_model.py (hunk 4 above)
Comment (sakshie95) [reply src/openjd/model/v2023_09/_model.py:511]: Ah well this is subjective I would rather teach the LLMs to follow the convention to keep the code clean :)) but non-blocking nonetheless, approved!

**Review by sakshie95 — APPROVED** [review]

**Review by jericht — APPROVED** [review]

File: test/openjd/model/test_parse.py (hunk 2 above)
Comment (mwiebe) [reply test/openjd/model/test_parse.py:205]: Good call, that was in there because of a different change that was fixed later.

File: src/openjd/model/v2023_09/_model.py [hunk 7]
```diff
@@ -2785,10 +3161,13 @@ def _permitted_extension_names(
...
     @field_validator("parameterDefinitions")
     @classmethod
-    def _unique_parameter_names(
-        cls, v: Optional[JobParameterDefinitionList]
+    def _validate_parameter_definitions(
+        cls, v: Optional[JobParameterDefinitionList], info: ValidationInfo
     ) -> Optional[JobParameterDefinitionList]:
         if v is not None:
+            # EnvironmentTemplate always has max 50 parameters (no extension increases this)
+            if len(v) > 50:
```
Comment (leongdl) [inline src/openjd/model/v2023_09/_model.py:3169]: Super nit - do we put constants like this somewhere ? Could be useful as an exported param?

File: src/openjd/model/v2023_09/_model.py (hunk 6 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:477]: Yeah, it's worthwhile describing that here. It's specified in https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#61-embeddedfiletext

You're right that it's not in https://github.com/OpenJobDescription/openjd-specifications/blob/mainline/rfcs/0004-enhanced-limits-and-capabilities.md. It was, and then we noted how clunky it was to include the block diffs in the RFC and shifted them out into the PR itself. But in doing so, we never added descriptions of those changes back into the RFC. I'm inclined to leave it as is and proceed with what we have in the spec, and watch for this phenomenon in future RFCs.

File: src/openjd/model/v2023_09/_model.py [hunk 8]
```diff
@@ -2605,10 +2973,18 @@ def _unique_step_names(cls, v: StepTemplateList) -> StepTemplateList:
...
     @classmethod
-    def _unique_parameter_names(
-        cls, v: Optional[JobParameterDefinitionList]
+    def _validate_parameter_definitions(
+        cls, v: Optional[JobParameterDefinitionList], info: ValidationInfo
     ) -> Optional[JobParameterDefinitionList]:
         if v is not None:
+            # Validate max length based on extension
+            context = cast(Optional[ModelParsingContext], info.context)
+            max_len = 200 if context and "FEATURE_BUNDLE_1" in context.extensions else 50
```
Comment (leongdl) [inline src/openjd/model/v2023_09/_model.py:2982]: The customer will be happy!

**Review by leongdl — APPROVED** [review]

File: src/openjd/model/v2023_09/_model.py (hunk 7 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:3169]: The source of truth for that is here: https://github.com/OpenJobDescription/openjd-specifications/wiki/2023-09-Template-Schemas#12-environment-template

The conformance review prompt that I showed in the PR description did note this value and that it matched the spec in the report it produced. I also think the conformance suite in https://github.com/OpenJobDescription/openjd-specifications/pull/103 is a good place documenting it. These feel sufficient to me, so I'd rather not add another version of these constants in another separate place.

File: src/openjd/model/v2023_09/_model.py (hunk 6 above)
Comment (mwiebe) [reply src/openjd/model/v2023_09/_model.py:477]: I'll merge this now, and then fix up your two comments in a trivial follow-up
